File: trading/cycle_quality.py
"""Display-only observations of an original two-order cycle batch.

Wall timestamps are Unix seconds and elapsed times are milliseconds. Prices,
quantities and spreads are decimal strings (36 significant digits for ratios).
Monotonic clock readings and market snapshot objects are never persisted.
"""
from decimal import Context, Decimal, localcontext
from fractions import Fraction
import math
import time

from .cycle import _depth_sweeps
from .exchange import RequestNotSent
from .models import positive, wire


def timestamp(value):
    return value if type(value) in (int, float) and math.isfinite(value) and value >= 0 else None


def number(value):
    value = Fraction(value)
    with localcontext(Context(prec=36)):
        result = format(Decimal(value.numerator) / Decimal(value.denominator), "f")
    return result.rstrip("0").rstrip(".") if "." in result else result


def estimate(quantity, depth=None, checked_at=None):
    result = {"status": "unavailable", "sampled_at": None, "checked_at": timestamp(checked_at),
              "quantity": quantity, "buy_vwap": None, "sell_vwap": None, "spread_bp": None}
    if depth is None:
        return result
    try:
        result["sampled_at"] = timestamp(depth.timestamp)
        checked_at = result["checked_at"] if result["checked_at"] is not None else time.time()
        bids, asks = _depth_sweeps(depth, checked_at)
        qty = Fraction(positive(quantity))
        buy, sell = asks.amount(qty) / qty, bids.amount(qty) / qty
        result.update(status="available", checked_at=checked_at, buy_vwap=number(buy),
                      sell_vwap=number(sell), spread_bp=number((buy - sell) * 20000 / (buy + sell)))
    except Exception:
        # The original admission checks decide whether orders may be sent.
        # Failed display observations must not introduce a new trading gate.
        pass
    return result


def new_quality(intent, trigger=None):
    trigger = trigger if isinstance(trigger, dict) else {}
    quantity = wire(positive(intent["quantity"]))
    source = trigger.get("source")
    return {"version": 1, "intent_id": intent.get("id"), "symbol": intent["symbol"], "phase": intent["phase"],
            "quantity": quantity, "created_at": timestamp(intent.get("created_at")), "updated_at": time.time(),
            "trigger": {"source": source if source in ("bbo", "depth", "poll") else "unknown",
                        "received_at": timestamp(trigger.get("received_at"))},
            "trigger_estimate": estimate(quantity, trigger.get("depth"), trigger.get("checked_at")),
            "final_estimate": estimate(quantity),
            "timing": {"request_status": "unknown", "request_started_at": None, "response_received_at": None,
                       "trigger_to_request_ms": None, "final_check_to_request_ms": None,
                       "request_to_response_ms": None},
            "actual": actual(intent)}


def actual(intent):
    result = {"status": "unknown", "confirmed": False, "quantity": None, "buy_vwap": None,
              "sell_vwap": None, "spread_bp": None, "buy_quantity": None, "sell_quantity": None,
              "buy_status": None, "sell_status": None, "repairs_present": bool(intent.get("repairs"))}
    try:
        # Read original orders only. A repair can flatten exposure but cannot
        # turn an incomplete original pair into a fully executed pair.
        from .execution import Executor, TERMINAL
        orders = intent["orders"]
        if len(orders) != 2 or {order["side"] for order in orders} != {"BUY", "SELL"} \
                or {order["positionSide"] for order in orders} != {"LONG", "SHORT"}:
            return result
        qty = Fraction(positive(intent["quantity"]))
        rows = []
        for order in orders:
            side = order["side"].lower()
            if order["symbol"] != intent["symbol"] or Fraction(positive(order["quantity"])) != qty:
                return result
            row = intent.get("receipts", {}).get(order["newClientOrderId"])
            if row is not None:
                Executor.validate_receipt(order, row)
                result[side + "_quantity"] = wire(positive(row["executedQty"], True))
                result[side + "_status"] = row["status"]
            rows.append(row)
        if all(row is not None and row["status"] == "FILLED" for row in rows):
            prices = {order["side"]: Fraction(positive(row["avgPrice"])) for order, row in zip(orders, rows)}
            buy, sell = prices["BUY"], prices["SELL"]
            result.update(status="filled", confirmed=True, quantity=wire(positive(intent["quantity"])),
                          buy_vwap=number(buy), sell_vwap=number(sell),
                          spread_bp=number((buy - sell) * 20000 / (buy + sell)))
        elif any(row is not None and positive(row["executedQty"], True) for row in rows):
            result["status"] = "partial"
        elif all(row is not None and row.get("local_not_sent") for row in rows):
            result["status"] = "not_sent"
        elif all(row is not None and row["status"] in TERMINAL for row in rows):
            result["status"] = "rejected"
        elif any(row is not None and row["status"] not in TERMINAL for row in rows):
            result["status"] = "pending"
    except Exception:
        # Invalid or incomplete receipts provide no evidence of a full pair.
        result.update(status="unknown", confirmed=False, quantity=None, buy_vwap=None,
                      sell_vwap=None, spread_bp=None)
    return result


def elapsed(start, end):
    start, end = timestamp(start), timestamp(end)
    if start is None or end is None or end < start:
        return None
    value = (end - start) * 1000
    return value if math.isfinite(value) else None


class ObservedBroker:
    """Per-call adapter: never replace or modify a shared broker method."""

    def __init__(self, broker, quality, trigger_ticks=None, final_ticks=None):
        self._broker, self._quality = broker, quality
        self._trigger_ticks, self._final_ticks = trigger_ticks, final_ticks

    def __getattr__(self, name):
        return getattr(self._broker, name)

    def submit(self, orders):
        started = None
        try:
            started = time.monotonic()
            timing = self._quality["timing"]
            timing.update(request_started_at=time.time(), request_status="unknown",
                          trigger_to_request_ms=elapsed(self._trigger_ticks, started),
                          final_check_to_request_ms=elapsed(self._final_ticks, started))
        except Exception:
            pass
        try:
            result = self._broker.submit(orders)
        except Exception as exc:
            try:
                self._quality["timing"]["request_status"] = "not_sent" if isinstance(exc, RequestNotSent) else "failed"
            except Exception:
                pass
            raise
        else:
            try:
                finished = time.monotonic()
                self._quality["timing"].update(request_status="returned", response_received_at=time.time(),
                                               request_to_response_ms=elapsed(started, finished))
            except Exception:
                pass
            return result

File: dashboard/lib/cycle-quality.ts
export type CycleExecutionEstimate = {
  status: 'available' | 'unavailable';
  sampled_at: number | null;
  checked_at: number | null;
  quantity: string | null;
  buy_vwap: string | null;
  sell_vwap: string | null;
  spread_bp: string | null;
};

export type CycleExecutionQuality = {
  version: 1;
  intent_id: string;
  symbol: string;
  phase: 'open' | 'close';
  quantity: string;
  created_at: number;
  updated_at: number;
  trigger: {
    source: 'bbo' | 'depth' | 'poll' | 'unknown';
    received_at: number | null;
  };
  trigger_estimate: CycleExecutionEstimate | null;
  final_estimate: CycleExecutionEstimate | null;
  timing: {
    request_started_at: number | null;
    response_received_at: number | null;
    trigger_to_request_ms: number | null;
    final_check_to_request_ms: number | null;
    request_to_response_ms: number | null;
    request_status: 'returned' | 'failed' | 'not_sent' | 'unknown';
    database?: {
      lock_wait_ms: number | null;
      connection_ms: number | null;
      connections: number;
    };
    transport?: {
      guard_ms?: number | null;
      budget_ms?: number | null;
      signing_ms?: number | null;
      before_http_ms?: number | null;
      http_ms?: number | null;
      response_decode_ms?: number | null;
      http_started_at?: number | null;
      http_finished_at?: number | null;
    };
    pre_submit?: {
      queue_ms: number | null;
      initial_account_ms: number | null;
      planning_ms: number | null;
      final_account_ms: number | null;
      final_check_ms: number | null;
      persist_ms: number | null;
      other_ms: number | null;
    };
  };
  actual: {
    status:
      | 'filled'
      | 'partial'
      | 'pending'
      | 'rejected'
      | 'not_sent'
      | 'unknown';
    confirmed: boolean;
    quantity: string | null;
    buy_vwap: string | null;
    sell_vwap: string | null;
    spread_bp: string | null;
    buy_quantity: string | null;
    sell_quantity: string | null;
    buy_status: string | null;
    sell_status: string | null;
    repairs_present: boolean;
  };
};

function text(value: unknown, fallback = '—'): string {
  return typeof value === 'string' && value.trim() ? value : fallback;
}

function mappedText(
  labels: Record<string, string>,
  value: unknown,
  fallback: string,
): string {
  return typeof value === 'string' && Object.hasOwn(labels, value)
    ? labels[value]
    : fallback;
}

// Preserve the service's decimal strings, including favorable negative spread.
// Quantity equality must not round small residuals through Number.
function decimal(value: unknown, signed = false, zero = true): string {
  if (typeof value !== 'string' || !/^-?(?:\d+(?:\.\d*)?|\.\d+)$/.test(value))
    return '—';
  if (!signed && value.startsWith('-')) return '—';
  if (!zero && /^-?(?:0+(?:\.0*)?|\.0+)$/.test(value)) return '—';
  return value;
}

function sameQuantity(left: unknown, right: unknown): boolean {
  if (
    decimal(left, false, false) === '—' ||
    decimal(right, false, false) === '—'
  )
    return false;
  const normalized = (value: string) => {
    const [whole, fraction = ''] = value.split('.');
    return `${whole.replace(/^0+/, '') || '0'}.${fraction.replace(/0+$/, '')}`;
  };
  return normalized(left as string) === normalized(right as string);
}

function validTime(value: unknown): value is number {
  return (
    typeof value === 'number' &&
    Number.isFinite(value) &&
    value >= 0 &&
    Number.isFinite(new Date(value * 1000).getTime())
  );
}

export function cycleQualityTime(value: number | null | undefined): string {
  return validTime(value)
    ? `${new Date(value * 1000).toISOString().replace('T', ' ').replace('Z', ' UTC')}`
    : '—';
}

function duration(value: unknown, from: unknown, to: unknown): string {
  return typeof value === 'number' &&
    Number.isFinite(value) &&
    value >= 0 &&
    validTime(from) &&
    validTime(to) &&
    to >= from
    ? String(value)
    : '—';
}

function estimateView(
  estimate: CycleExecutionEstimate | null | undefined,
  quantity: string,
) {
  const same = sameQuantity(estimate?.quantity, quantity);
  const spread = decimal(estimate?.spread_bp, true);
  const buy = decimal(estimate?.buy_vwap, false, false);
  const sell = decimal(estimate?.sell_vwap, false, false);
  const available =
    estimate?.status === 'available' &&
    same &&
    spread !== '—' &&
    buy !== '—' &&
    sell !== '—';
  const status = !estimate
    ? '未提供'
    : estimate.status === 'unavailable'
      ? '估计不可用'
      : !same
        ? '数量缺失或不一致'
        : !available
          ? '估计数据不完整'
          : '同数量估计';
  return {
    available,
    status,
    spread: available ? spread : '—',
    quantity: decimal(estimate?.quantity, false, false),
    buy,
    sell,
    sampledAt: cycleQualityTime(estimate?.sampled_at),
    checkedAt: cycleQualityTime(estimate?.checked_at),
  };
}

const ACTUAL_STATUS: Record<string, string> = {
  filled: '双边已确认',
  partial: '部分成交',
  pending: '待确认',
  rejected: '已拒绝',
  not_sent: '未发单',
  unknown: '未确认',
};
const ORDER_STATUS: Record<string, string> = {
  FILLED: '全部成交',
  PARTIALLY_FILLED: '部分成交',
  NEW: '已接受，待成交',
  CANCELED: '已撤销',
  EXPIRED: '已过期',
  EXPIRED_IN_MATCH: '撮合时失效',
  REJECTED: '已拒绝',
};
const REQUEST_STATUS: Record<string, string> = {
  returned: '调用已返回',
  failed: '调用失败，响应未确认',
  not_sent: '未发单',
  unknown: '未记录',
};

const PRE_SUBMIT_STAGES = [
  ['queue_ms', '触发→账户任务开始'],
  ['initial_account_ms', '账户热数据读取'],
  ['planning_ms', '首轮规划与盘口'],
  ['final_account_ms', '提交前账户复核'],
  ['final_check_ms', '最终盘口与风控检查'],
  ['persist_ms', '订单记录写入'],
  ['other_ms', '其他本地准备'],
] as const;

export function cycleExecutionQualityView(
  quality?: CycleExecutionQuality | null,
) {
  if (!quality || quality.version !== 1) return null;
  const quantity = decimal(quality.quantity, false, false);
  const actual = quality.actual;
  const spread = decimal(actual?.spread_bp, true);
  const buy = decimal(actual?.buy_vwap, false, false);
  const sell = decimal(actual?.sell_vwap, false, false);
  const same =
    sameQuantity(actual?.quantity, quantity) &&
    sameQuantity(actual?.buy_quantity, quantity) &&
    sameQuantity(actual?.sell_quantity, quantity);
  const filled =
    actual?.status === 'filled' &&
    actual.confirmed === true &&
    actual.buy_status === 'FILLED' &&
    actual.sell_status === 'FILLED';
  const complete =
    filled && same && spread !== '—' && buy !== '—' && sell !== '—';
  const actualStatus =
    actual?.status !== 'filled'
      ? mappedText(ACTUAL_STATUS, actual?.status, '成交状态未记录')
      : !filled
        ? '成交尚未完整确认'
        : !same
          ? '数量缺失或不一致'
          : !complete
            ? '成交数据不完整'
            : ACTUAL_STATUS.filled;
  const triggerSource = quality.trigger?.source;
  const wsTrigger = triggerSource === 'bbo' || triggerSource === 'depth';
  const knownTrigger = wsTrigger || triggerSource === 'poll';
  const timing = quality.timing;
  return {
    intentId: text(quality.intent_id),
    symbol: text(quality.symbol, '品种未记录'),
    phase:
      quality.phase === 'open'
        ? '开仓'
        : quality.phase === 'close'
          ? '平仓'
          : '阶段未记录',
    quantity,
    updatedAt: cycleQualityTime(quality.updated_at),
    createdAt: cycleQualityTime(quality.created_at),
    trigger: estimateView(quality.trigger_estimate, quantity),
    final: estimateView(quality.final_estimate, quantity),
    actual: {
      complete,
      status: actualStatus,
      spread: complete ? spread : '—',
      buy: filled && same ? buy : '—',
      sell: filled && same ? sell : '—',
      quantity: filled && same ? decimal(actual?.quantity, false, false) : '—',
      buyQuantity: decimal(actual?.buy_quantity),
      sellQuantity: decimal(actual?.sell_quantity),
      buyStatus: mappedText(
        ORDER_STATUS,
        actual?.buy_status,
        text(actual?.buy_status, '未确认'),
      ),
      sellStatus: mappedText(
        ORDER_STATUS,
        actual?.sell_status,
        text(actual?.sell_status, '未确认'),
      ),
      repairs: actual?.repairs_present === true,
    },
    timing: [
      {
        label:
          triggerSource === 'poll'
            ? '轮询唤醒 → 请求开始'
            : 'WS 接收 → 请求开始',
        value: knownTrigger
          ? duration(
              timing?.trigger_to_request_ms,
              quality.trigger?.received_at,
              timing?.request_started_at,
            )
          : '—',
      },
      {
        label: '最终深度检查 → 请求开始',
        value: duration(
          timing?.final_check_to_request_ms,
          quality.final_estimate?.checked_at,
          timing?.request_started_at,
        ),
      },
      {
        label: '请求开始 → 收到响应',
        value:
          timing?.request_status === 'returned'
            ? duration(
                timing.request_to_response_ms,
                timing.request_started_at,
                timing.response_received_at,
              )
            : '—',
      },
    ],
    preSubmit: PRE_SUBMIT_STAGES.map(([key, label]) => {
      const value = timing?.pre_submit?.[key];
      return {
        key,
        label,
        value:
          typeof value === 'number' && Number.isFinite(value) && value >= 0
            ? String(value)
            : '—',
      };
    }),
    database: [
      ['lock_wait_ms', '等待账本访问锁'],
      ['connection_ms', '建立数据库连接'],
    ].map(([key, label]) => {
      const value = timing?.database?.[key as 'lock_wait_ms' | 'connection_ms'];
      return { label, value: observedDuration(value) };
    }),
    transport: [
      ['guard_ms', '发单前本地校验'],
      ['budget_ms', '请求额度检查'],
      ['signing_ms', '签名与请求编码'],
      ['http_ms', 'HTTP 调用（含连接处理）'],
      ['response_decode_ms', '解析响应'],
    ].map(([key, label]) => ({
      label,
      value: observedDuration(
        timing?.transport?.[
          key as keyof NonNullable<CycleExecutionQuality['timing']['transport']>
        ],
      ),
    })),
    beforeHttp: observedDuration(timing?.transport?.before_http_ms),
    triggerSource:
      triggerSource === 'bbo'
        ? 'WS 最优盘口'
        : triggerSource === 'depth'
          ? 'WS 深度'
          : triggerSource === 'poll'
            ? '轮询唤醒'
            : '触发来源未记录',
    triggerReceivedAt: cycleQualityTime(quality.trigger?.received_at),
    requestStartedAt: cycleQualityTime(timing?.request_started_at),
    responseReceivedAt:
      timing?.request_status === 'returned'
        ? cycleQualityTime(timing.response_received_at)
        : '—',
    requestStatus: mappedText(REQUEST_STATUS, timing?.request_status, '未记录'),
  };
}

function observedDuration(value: unknown) {
  return typeof value === 'number' && Number.isFinite(value) && value >= 0
    ? String(value)
    : '—';
}
